pkg/controller/hostpathprovisioner/scc.go

```diff
@@ -0,0 +1,186 @@
+	setLastAppliedConfiguration(desiredMetaObj)
+
+	// Set HostPathProvisioner instance as the owner and controller
+	if err := controllerutil.SetControllerReference(cr, desired, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if this SecurityContextConstraints already exists
+	found := &secv1.SecurityContextConstraints{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: cr.Name}, found)
+	if err != nil && errors.IsNotFound(err) {
+		reqLogger.Info("Creating a new SecurityContextConstraints", "SecurityContextConstraints.Name", desired.Name)
+		err = r.client.Create(context.TODO(), desired)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// SecurityContextConstraints created successfully - don't requeue
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Keep a copy of the original for comparison later.
+	currentRuntimeObjCopy := found.DeepCopyObject()
+
+	// allow users to add new annotations (but not change ours)
+	mergeLabelsAndAnnotations(desiredMetaObj, &found.ObjectMeta)
+
+	// create merged SecurityContextConstraints from found and desired.
+	merged, err := mergeObject(desired, found)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// SecurityContextConstraints already exists, check if we need to update.
+	if !reflect.DeepEqual(currentRuntimeObjCopy, merged) {
+		logJSONDiff(reqLogger, currentRuntimeObjCopy, merged)
+		// Current is different from desired, update.
+		reqLogger.Info("Updating SecurityContextConstraints", "SecurityContextConstraints.Name", desired.Name)
+		err = r.client.Update(context.TODO(), merged)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+	// SecurityContextConstraints already exists and matches the desired state - don't requeue
+	reqLogger.Info("Skip reconcile: SecurityContextConstraints already exists", "SecurityContextConstraints.Name", found.Name)
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileHostPathProvisioner) deleteSCC(cr *hostpathprovisionerv1alpha1.HostPathProvisioner) error {
+	if used, err := r.checkSCCUsed(); used == false {
+		return err
+	}
+	// Check if this SecurityContextConstraints already exists
+	scc := &secv1.SecurityContextConstraints{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: cr.Name}, scc)
+	if err != nil && errors.IsNotFound(err) {
+		// Already gone, return
+		return nil
+	} else if err != nil {
+		return err
+	}
+	// Delete SCC
+	return r.client.Delete(context.TODO(), scc)
+}
+
+func (r *ReconcileHostPathProvisioner) addSCCFinalizer(reqLogger logr.Logger, cr *hostpathprovisionerv1alpha1.HostPathProvisioner) error {
+	if used, err := r.checkSCCUsed(); used == false {
+		return err
+	}
+	if len(cr.GetFinalizers()) < 1 && cr.GetDeletionTimestamp() == nil {
+		reqLogger.Info("Adding SCC Finalizer")
+		cr.SetFinalizers([]string{"finalizer.scc.hostpath-provisioner"})
```

Comment:
Why is this needed?

Reply:
So even though the operator controller owns the scc, when the CR is removed, it didn't remove the scc automatically as well. I have a feeling it is an Open Shift protection thing. So I had to use a finalizer to specifically remove the SCC when the CR is removed.